maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java

```diff
@@ -3623,9 +3623,20 @@ public void setSystemProperties( Properties systemProperties )
     }
 
     @SuppressWarnings( "UnusedDeclaration" )
-    public void setSystemPropertyVariables( Map<String, String> systemPropertyVariables )
+    public void setSystemPropertyVariables( Map<String, ?> systemPropertyVariables )
     {
-        this.systemPropertyVariables = systemPropertyVariables;
+        if (systemPropertyVariables != null)
+        {
+            this.systemPropertyVariables = new HashMap<>();
+            for ( final Map.Entry<String, ?> entry : systemPropertyVariables.entrySet() )
+            {
+                this.systemPropertyVariables.put( entry.getKey(), String.valueOf( entry.getValue() ) );
```

Comment:
See how the getter is used and how the null check looks like. I think it is something like `val == null ? "" : val`. It would be nice to extract this treatment or even more code to a private static method and call it from both places.

Reply:
@rmannibucau 
I was talking [about](https://github.com/apache/maven-surefire/blob/master/maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireProperties.java#L175).
Due to the method [copyProperties](https://github.com/apache/maven-surefire/blob/master/maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireProperties.java#L168) is called only in [calculateEffectiveProperties](https://github.com/apache/maven-surefire/blob/master/maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireProperties.java#L148) which is called in ASM, then the solution is to move both methods to ASM because they are used only there, and then your method can call [copyProperties](https://github.com/apache/maven-surefire/blob/master/maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireProperties.java#L168). You have to only change `Properties` to `Map` withou generics. The method `copyProperties` takes your `systemPropertyVariables` to the second parameter.

Reply:
Yep got it but have to admit I'm mixed about it, I like your setter proposal because it makes the abstract data consistent for any consumer whereas copyProperties can let subclasses/consumers see wrong data which is worse IMHO.
About the utility I'm tempted to say the gain is very light and mainly envisionned because we can't use lambda yet. Since maven master is on java 8 and plugins already started to move I suspect this will drop as soon as we move to java 8 so let's keep it simple maybe?
Wdyt?

Reply:
Both methods will be private and they have a commpon purpose - moving properties from one Map to second Map. Of course j8 may turn loops to streams but that's a big hammer for this little refactoring.
I would be glad when Maven aligns all default Maven plugins to j8 after Maven 4.0, and then all the stack would be on the same j8. IMO j9 would be even a bigger benefit for the Maven but not for users yet ;-).

Reply:
Understood but it does not fix the issue that the mojo exposed view is wrong for some time in between until the utility is public which is worse than these 2 lines IMHO so not a bad compromise probably.

Reply:
Even if the utility class is public, still the method `calculateEffectiveProperties` is package private, and `copyProperties` is private.

Reply:
Right, but once again it is not the main reason to not fix it there as explained.

Reply:
I don't know what you mean since we said many things.
j8 is in the project due to the JUnit5 but the compiler ha target j7, and all the [core plugins](https://maven.apache.org/plugins/) are at j7. We all know that there will be one day when Maven and the core plugins will be alligned with the same java version.
IMO the refactiring where you **reuse and safely move** an important method [copyProperties](https://github.com/apache/maven-surefire/blob/master/maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireProperties.java#L168) to your place where you need to have the fix can be freely done as a good fix.

Reply:
About java 8: it is maven 4 time since it is already set to 8 but not v3, agree.
About why setter is saner than any other "flow" method: because class is public+abstract and getter as well so it is part of the contract of the class to convert it in the setter - as you said on slack - and not later on in the flow which can be missed by the contract. So we actually don't have much choice. We can create an utility but as mentionned I think it is worse than duplicating the inline test IMHO so think we are good.
Now if you want to fix it differently I'll not fight while it behaves the same.

side note: i'll not get time this week end/start of next week to modify this PR and can be slow to answer.